scripts/data/src/org/akvo/gae/remoteapi/FixOrphanedSubmissions.java

```diff
@@ -0,0 +1,152 @@
+                doIt = true;
+                break;
+            }
+        }
+        String pathname = "/tmp/" + instanceName + "-orphaned-instances.csv";
+        final File f = new File(pathname);
+
+        FileUtils.write(f, "FormId,DatapointIdentifier,OrphanedMonitoringInstanceId\n");
+        List<Long> surveyGroupIds = getMonitoringSurveys(ds);
+        System.out.printf("Found %d monitoring surveys\n", surveyGroupIds.size());
+        List<Entity> dataPoints = getDataPoints(ds, surveyGroupIds);
+        System.out.printf("Found %d monitoring datapoints\n", dataPoints.size());
+
+        Set<Long> dataPointIds = dataPoints
+                .stream()
+                .map(entity -> entity.getKey().getId())
+                .collect(Collectors.toSet());
+
+        Map<Long, Set<Long>> mappedBySurveyId = new HashMap<>();
+        for (Entity entity: dataPoints) {
+            Long surveyId = (Long) entity.getProperty("creationSurveyId");
+            Long dataPointId = entity.getKey().getId();
+            if (mappedBySurveyId.get(surveyId) == null) {
+                Set<Long> datapoints = new HashSet<>();
+                datapoints.add(dataPointId);
+                mappedBySurveyId.put(surveyId, datapoints);
+            } else {
+                mappedBySurveyId.get(surveyId).add(dataPointId);
+            }
+        }
+
+        Set<Long> dataPointsFound = fetchExistingDataPointInstances(ds, mappedBySurveyId);
+        dataPointIds.removeAll(dataPointsFound);
+
+        List<Entity> instances = getSurveyInstances(ds, new ArrayList<>(dataPointIds));
+
+        final StringBuilder sb = new StringBuilder();
+        if (instances != null) {
+            for (Entity instance : instances) {
+                Key key = instance.getKey();
+                surveyInstances.add(key);
+                sb.append(
+                        String.format("%s,%s,%s", instance.getProperty("surveyId"), instance.getProperty("surveyedLocaleIdentifier"), key.getId()))
+                        .append("\n");
+            }
+            FileUtils.write(f, sb.toString(), true);
+        }
+        if (surveyInstances.size() > 0) {
+            System.out.printf("Found a total of %d orphaned instances\n", surveyInstances.size());
+            if (doIt) {
+                batchDelete(ds, surveyInstances);
+            }
+        } else {
+            System.out.println("No orphaned SurveyInstances found");
+        }
+    }
+
+    private Set<Long> fetchExistingDataPointInstances(DatastoreService ds, Map<Long, Set<Long>> mappedBySurveyId) {
+        Set<Long> dataPointIds = new HashSet<>();
+        Set<Entity> allDataPoints = new HashSet<>();
+        for (Long surveyId : mappedBySurveyId.keySet()) {
+            Query.Filter f = new Query.FilterPredicate("surveyId", Query.FilterOperator.EQUAL, surveyId);
+            Query.Filter f2 = new Query.FilterPredicate("surveyedLocaleId", Query.FilterOperator.IN, mappedBySurveyId.get(surveyId));
```

Comment:
I think in this case we probably do not need to filter by the datapoint Ids that have been retrieved. We want all the form instances that are present for an existing registration form form, and then extract a set of datapoint ids from them. so that we can compare if indeed for every form instance that exists, there is also a datapoint.